### PR TITLE
Follow recommended blog from "recommended_blogs" discovery card

### PR DESCRIPTION
**Pull request by AdamGrzybkowski**
Initial draft PR to discuss the code 

Questions:
1. Should we show "enable notification" snack when following the Site from the "recommended_blogs" card?

To test:
- open Discovery tab
- scroll to 'Sites to follow' card
- click follow/unfollow IconButton
- verify site was followed/unfollowed

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderBlogTableWrapper.kt [hunk 1]
```diff
@@ -11,4 +11,13 @@ class ReaderBlogTableWrapper
     fun isNotificationsEnabled(blogId: Long): Boolean = ReaderBlogTable.isNotificationsEnabled(blogId)
     fun setNotificationsEnabledByBlogId(blogId: Long, isEnabled: Boolean) =
             ReaderBlogTable.setNotificationsEnabledByBlogId(blogId, isEnabled)
+    fun isSiteFollowed(blogId: Long, feedId: Long?): Boolean {
+        return if (blogId != 0L) {
+            ReaderBlogTable.isFollowedBlog(blogId)
+        } else if (feedId != null && feedId != 0L) {
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderBlogTableWrapper.kt]: I did add `feedId != 0L` check for compatibility reasons. It shouldn't be needed with the new code where `feedId` is nullable.

File: WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+data class RecommendedBlog(
+    val name: String,
+    val url: String,
+    val blogId: Long,
+    val description: String,
+    val iconUrl: String?,
+    val feedId: Long?,
+    val isFollowed: Boolean
+)
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: I've decided to create a new model that will combine information from two sources:
1. Discovery endpoint - blog details
2. Blog Table - whether a blog is followed or not

I don't want to store `isFollowed` information alongside recommended_blogs card because this would create another source of truth instead of having one (Blog table). I believe this is done like that with the Posts and this complicates the logic when following the Site. Take a look [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderBlogActions.java#L197). If I would follow this pattern we would have to add third line to store the same information. 

So with the current implementation the data flow would look like this:
1. Fetch `recommended_blogs` 
2. Store it locally - currently this is stored as plain json, so there's no info about the follow state
3. Get Discovery Cards -> here we retrieve the recommended blogs and map it to the new model where we additionally check the information whether a site is followed or not in the Blog table

thanks to this when we follow a Site we don't have to store it in a separate places.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 3]
```diff
@@ -1614,35 +1613,6 @@ public void onEventMainThread(ReaderEvents.SearchPostsEnded event) {
         }
     }
 
-    /*
-     * called when user taps follow item in popup menu for a post
-     */
-    private void toggleFollowStatusForPost(final ReaderPost post) {
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:1616]: I would have to update this `ReaderBlogActions.followBlogForPost` but this whole method was unused so I simply removed it

File: WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderBlogActions.java [hunk 4]
```diff
@@ -239,18 +239,14 @@ public void onErrorResponse(VolleyError volleyError) {
...
-                                            boolean isAskingToFollow,
-                                            ActionListener actionListener) {
-        if (post == null) {
-            AppLog.w(T.READER, "follow action performed with null post");
-            ReaderActions.callActionListener(actionListener, false);
-            return false;
-        }
-        if (post.feedId != 0) {
-            return followFeedById(post.feedId, isAskingToFollow, actionListener);
+    public static boolean followBlog(final Long blogId,
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderBlogActions.java:242]: Refactored and Renamed it to decouple it from the `Post`

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverAdapter.kt [hunk 5]
```diff
@@ -75,8 +75,11 @@ class ReaderDiscoverAdapter(
                 is ReaderPostUiState -> {
                     oldItem.postId == (newItem as ReaderPostUiState).postId && oldItem.blogId == newItem.blogId
                 }
+                is ReaderRecommendedBlogsCardUiState -> {
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverAdapter.kt:78]: I had to update this logic to avoid item animation when `isFollowed` property is changed

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt [hunk 6]
```diff
@@ -141,8 +141,44 @@ class ReaderPostCardActionsHandler @Inject constructor(
         }
     }
 
+    suspend fun handleFollowRecommendedSiteClicked(blogId: Long, feedId: Long?, isAskingToFollow: Boolean) {
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt]: For now, I did copy-paste the code from another function but I think this could be unified

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt [hunk 7]
```diff
@@ -141,8 +141,44 @@ class ReaderPostCardActionsHandler @Inject constructor(
...
+                    _followStatusUpdated.postValue(it)
+                    siteNotificationsUseCase.fetchSubscriptions()
+
+//                    if (it.showEnableNotification) {
+//                        val action = prepareEnableNotificationSnackbarAction(post.blogName, post.blogId)
+//                        action.invoke()
+//                    } else if (it.deleteNotificationSubscription) {
+//                        siteNotificationsUseCase.updateSubscription(it.blogId, DELETE)
+//                        siteNotificationsUseCase.updateNotificationEnabledForBlogInDb(it.blogId, false)
+//                    }
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt]: Should we show the same info when following the Site from the "recommended_blogs" card?

File: WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/GetDiscoverCardsUseCase.kt [hunk 8]
```diff
@@ -63,6 +66,20 @@ class GetDiscoverCardsUseCase @Inject constructor(
...
                                     val recommendedBlogs = parseDiscoverCardsJsonUseCase.parseRecommendedBlogsCard(it)
+                                            .map { blog ->
+                                                RecommendedBlog(
+                                                        blogId = blog.blogId,
+                                                        url = blog.url,
+                                                        name = blog.name,
+                                                        iconUrl = blog.iconUrl,
+                                                        feedId = blog.feedId,
+                                                        description = blog.description,
+                                                        isFollowed = readerBlogTableWrapper.isSiteFollowed(
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/GetDiscoverCardsUseCase.kt]: This is the place where combining the info from two sources happens mentioned in the comment about new model

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt [hunk 9]
```diff
@@ -131,8 +130,8 @@ class ReaderDiscoverLogic(
         val fullCardsJson = json.optJSONArray(JSON_CARDS)
 
         // Parse the json into cards model objects
-        val cards = parseCards(fullCardsJson)
-        insertPostsIntoDb(cards.filterIsInstance<ReaderPostCard>().map { it.post })
+        val postCards = parsePostCards(fullCardsJson)
+        insertPostsIntoDb(postCards.map { it.post })
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt]: I've decided to rename the function and modify it to return only `ReaderPostCard` objects. 
It did confuse me before and other types are not used at all here so the code was unnecessary.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt [hunk 10]
```diff
@@ -189,9 +178,16 @@ class ReaderDiscoverLogic(
...
+                    val recommendedBlogs = parseDiscoverCardsJsonUseCase.parseRecommendedBlogsCard(cardJson)
+                    val recommendedBlogsJson = cardJson.optJSONArray(JSON_CARD_DATA)
+                    val notFollowedRecommencedBlogsJson = JSONArray()
+                    recommendedBlogs.forEachIndexed { blogIndex, blog ->
+                        if (!readerBlogTableWrapper.isSiteFollowed(blog.blogId, blog.feedId)) {
+                            notFollowedRecommencedBlogsJson.put(recommendedBlogsJson.get(blogIndex))
+                        }
+                    }
+                    if (notFollowedRecommencedBlogsJson.length() > 0) {
+                        simplifiedJson.put(index++, cardJson.put(JSON_CARD_DATA, notFollowedRecommencedBlogsJson))
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt]: In the previous PR, I used this place to filter out cards with an empty list. In this PR I've added extra logic to filter out sites that are followed. As I wrote in the commit message
> It is possible that API returns already followed sites. We should filter out such sites as it doesn't make sens to recommend a site that is already known to the user. This might get fixed in the API but for now we should do this in the app.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderSiteFollowUseCase.kt [hunk 11]
```diff
@@ -83,4 +97,21 @@ class ReaderSiteFollowUseCase @Inject constructor(
...
+            override val blogId: Long = post.blogId
+            override val feedId: Long = post.feedId
+        }
+
+        data class RecommendedSite(
+            override val blogId: Long,
+            override val feedId: Long,
+            val isAskingToFollow: Boolean
+        ) : Param()
+    }
```
Comment (AdamGrzybkowski) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderSiteFollowUseCase.kt]: In order to reuse this use case, I decided to create a sealed class for a param object that is passed to the `toggleFollow` function. 
This allows us to pass different properties and based on the type map it to the correct data.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt (hunk 10 above)
Comment (AdamGrzybkowski) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/services/discover/ReaderDiscoverLogic.kt]: I will remove this as per [this comment](https://wp.me/pbMoDN-14X#comment-2341)

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt (hunk 6 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt]: Yep, I think they can be unified.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt (hunk 7 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt]: Comparing with "follow site" functionality it seems we should. Can you add it as an open question in the PR description so that we remember revisiting it?

File: WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt (hunk 2 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: > I believe this is done like that with the Posts and this complicates the logic when following the Site

Can you clarify this a bit? And what difference in pattern do you see between following a post site (using more menu on the top right of the discover card) vs following a recommended site.

File: WordPress/src/main/res/layout/reader_recommended_blog_item.xml [hunk 12]
```diff
@@ -6,6 +6,19 @@
     android:layout_height="wrap_content"
     android:background="?attr/selectableItemBackground">
 
+    <androidx.appcompat.widget.AppCompatImageButton
+        android:id="@+id/site_follow_icon"
```
Comment (ashiagr) [inline WordPress/src/main/res/layout/reader_recommended_blog_item.xml]: Wondering if we can reuse `ReaderFollowButton` with `wpShowFollowButtonCaption` style here 🤔.

File: WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt (hunk 2 above)
Comment (AdamGrzybkowski) [reply WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: Sure! 
> And what difference in pattern do you see between following a post site (using more menu on the top right of the discover card) vs following a recommended site.

I think there should be no difference. It shouldn't matter where the `follow` button is clicked. That's why I did try to decouple it from `post` a bit by changing the params and function names in the code. 

The comment above was about the internal storing logic. Currently, whether a blog is followed or not is stored in two tables: `blog` & `post`. When you `follow/unfollow` a blog you need to update both tables. I would guess that [doing this for each post](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java#L756) stored locally comes with a small performance penalty. 

The approach I took here was to treat `Blog` table as the single source of truth for information about `follow` state. So instead of keeping `isFollow` locally in another table where recommended blogs are stored I combining this two information later on. 
This could also be done with Posts. Each Post object instead of keeping `isFollow` info could just keep a foreign key to the blog object. When loading posts we could build the full object with `isFollow` info by retrieving it from the `Blog` table. Thanks to this we would only have to store `isFollow` in one table.

Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: Thanks for the explanation :). I still have few questions regarding the original comment:

> I don't want to store isFollowed information alongside recommended_blogs card

> we would have to add third line to store the same information.

Haven’t we already created two sources for blogs data by saving recommended blogs json data into `tbl_discover_cards`?

I see that `score` field is returned in the json that cannot be inserted into `tbl_blog_info`. Similarly there are some missing fields from `tbl_blog_info` not being returned from the API for the `recommended_blog` card. 

Considering we’ve just few boolean `blog` fields that are missing from the `recommended_blog` response, if we receive these missing `blog` fields and ignore `score` field, how do you feel about keeping single data source for blogs by saving `recommended_blog` to `tbl_blog_info`? That ways we can reuse existing logic to update follow state and need not add a "third line to store the same information" or create new models.

I'm just trying to understand possibilities, feel free to put your arguments.

> Each Post object instead of keeping isFollow info could just keep a foreign key to the blog object. 

Totally agree. Let's keep it out of scope of this task at the moment :).

Comment (AdamGrzybkowski) [reply WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: > Haven’t we already created two sources for blogs data by saving recommended blogs json data into tbl_discover_cards?

Kinda yes, this may be the lack of knowledge of the project on my side but I thought only followed blogs are stored locally and I treated the "recommended blog" as a different thing. Additionally, we still want to store `recommended_card` separately as I believe the Discovery view is driven by the local storage.

> I see that score field is returned in the json that cannot be inserted into tbl_blog_info.

Yeah, but I'm not parsing that field as it's not needed, so I don't think that would be the problem for now. 

> Considering we’ve just few boolean blog fields that are missing from the recommended_blog response, if we receive these missing blog fields and ignore score field, how do you feel about keeping single data source for blogs by saving recommended_blog to tbl_blog_info?

Yeah I think we could do that as well. We would still need to store some information about the `recommended_blogs` card though.

> That ways we can reuse existing logic to update follow state and need not add a "third line to store the same information" or create new models.

We can reuse the current logic now? At least that's what I did that and it seems to work, or am I missing something?


So to sum up, I would say that we can use the  `tbl_blog_info` as long as we are not treating the recommended blog as a different thing and we are fine with storing a lot of blogs over time. 
If I was to make that decision(with my current knowledge about the project) I would probably stay with the current implementation as it keeps `recommended_blogs` in a separate place but it still combines `isFollow` information from the `tbl_blog_info` so no need for extra logic. 

One thing that I saw right now when writing this comment and going through the code is that there's `tbl_recommended_blogs` table but this one seems to be used specifically for one view as it drops all records when updating new ones

Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/models/discover/RecommendedBlog.kt]: > We can reuse the current logic now? At least that's what I did that and it seems to work, or am I missing something?

My comment was meant for this part of your answer:
> I believe this is done like that with the Posts and this complicates the logic when following the Site. 

If it shouldn't matter where the follow button is clicked, logic should not get complicated when following the Site.

Let’s continue with the discussion once you’ve created a PR in the main repo :).
